### PR TITLE
accounts/abi/bind: Add non-sending ABI methods

### PR DESCRIPTION
**Pull request by mcdee**
This patch creates additional methods for creating, and creating and signing, but not sending ABI methods.

To give an example of the ERC-20 function `transfer()` `abigen` currently creates a `Transfer()` method.  This patch creates in addition `CreateTransfer()` and `CreateAndSignTransfer()` methods to create, and create and sign, respectively a `Transfer()` transaction. Both new methods return the transaction, and neither attempt to send it.

This change is backwards-compatible by leaving the functionality of `Tranfser()` as-is.

### CODE REVIEW

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: ff6743d0c359410fb8489d4c22e27436d285f504
  * Commit subjects should be kept under 100 characters

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by mcdee** [comment]
@fjl does this PR address your request to have non-sending ABI calls?

**Comment by karalabe** [comment]
I've been thinking about this a bit, and I don't think we actually need new methods at all. All the currently generated methods return a transaction object. The only thing we need to control is whether that transaction object is a) raw unsigned b) signer unsubmitted c) signed and submitted.

There's no need to generate separate methods for these though as we already have a `TransactOpts` passed to each of the current arguments. As such, we can just massage it into the existing design:

 * Instead of having `TransactOpts.Signer` mandatory, we can make it optional, the omission of which would result in not signing and submitting the transaction at all, just returning the raw unsigned one.
 * We can add an additional `TransactOpts.Offline` field, the presence of which would mean that the transaction **is** signed, but it's returned to the user for submission through whatever different means.

Both these additions would keep the current behavior unchanged, but thy would add the desired functionality in a much more subtle and less invasive way.

**Comment by fjl** [comment]
I'm on board with making TransactOpts.Signer optional. Less sure about the Offline option, maybe we don't need it.

**Comment by fjl** [comment]
Let's make Signer optional first.

**Comment by mcdee** [comment]
@fjl is there another way you propose to obtain a signed but unsent transaction?

**Comment by fjl** [comment]
@mcdee No. I'm on board with @karalabe suggestion to make Signer optional in TransactOpts.

**Comment by karalabe** [comment]
@fjl I think @mcdee 's question was whether you have a particular proposal for "sign the transaction, but don't submit" / *offline mode*.

The optional signer will not sign the transaction at all, just create it. But there could be a use case to sign it also (e.g. offline machine), and then submit it from a different machine.

**Comment by fjl** [comment]
Signing transactions is easy, you can do that in any way once you have the transaction. I don't see why the binding should offer sign-but-not-send.

**Comment by mcdee** [comment]
@fji it's just as easy to send a transaction, but the code allows that.

The general flow of a transaction is create->sign->send.  It seems churlish to allow create or create->sign->send but not create->sign, and even though it's relatively easy it's still a step where a user could make mistakes.  For the sake of an option in `TransactOpts` it seems reasonable to provide exit at any of the three points in the flow.

**Comment by mcdee** [comment]
I've changed the patch to make `TransactOpts.Signer` optional.  `transact()` now has the following flow:

  - create the transaction
  - if `TransactOpts.Signer` is nil then return the unsigned transaction
  - sign the transaction
  - if `TransactOpts.DoNotSend` is true then return the signed transaction
  - send the transaction

**Comment by fabdarice** [comment]
I truly believe that this patch is very valuable.
The other way around to "create->sign->send" today is through `types.NewTransaction` and `types.SignTx`.

It's easy enough when dealing with a direct transfer.
However, calling a contract function requires to manually encode the function and inputs parameters and pass it as the data of `types.NewTransaction`.

This PR would make things a lot more streamlined and trivial.

File: accounts/abi/bind/base.go [hunk 1]
```diff
@@ -205,20 +205,27 @@ func (c *BoundContract) transact(opts *TransactOpts, contract *common.Address, i
...
+	// Create the transaction
 	var rawTx *types.Transaction
 	if contract == nil {
 		rawTx = types.NewContractCreation(nonce, value, gasLimit, gasPrice, input)
 	} else {
 		rawTx = types.NewTransaction(nonce, c.address, value, gasLimit, gasPrice, input)
 	}
 	if opts.Signer == nil {
-		return nil, errors.New("no signer to authorize the transaction with")
+		return rawTx, nil
```
[the remark this answers is not in the document]
Comment (mcdee) [reply accounts/abi/bind/base.go]: The process is create->sign->send.  If the signer is not present then there is no signed transaction to send so the state of `DoNotSend` is irrelevant.

File: accounts/abi/bind/base.go [hunk 2]
```diff
@@ -205,20 +205,27 @@ func (c *BoundContract) transact(opts *TransactOpts, contract *common.Address, i
...
 	}
+
+	// Sign the transaction
 	signedTx, err := opts.Signer(types.HomesteadSigner{}, opts.From, rawTx)
 	if err != nil {
 		return nil, err
 	}
+	if opts.DoNotSend {
```
[the remark this answers is not in the document]
Comment (mcdee) [reply accounts/abi/bind/base.go]: Is there any code benefit to this change?  In terms of human parsing this introduces a double negative, which doesn't seem like a good idea.

**Comment by mcdee** [comment]
@gballet this change is related to golang code generated from an ABI definition.  It has nothing to do with clef and is totally separate functionality.

As for testing, as far as I am aware there are no tests at all for this file so there is no base on which to build

File: accounts/abi/bind/base.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (gballet) [reply accounts/abi/bind/base.go]: *Your* process is create > sign > send. You don't know who else out there is expecting for that error that is no longer returned with this change. If they sync, `DoNotSend` will be false and the behavior should not change.

File: accounts/abi/bind/base.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (gballet) [reply accounts/abi/bind/base.go]: image

:laughing: 

The intent is to remove the number of exit points in the function that return the exact same thing, and which is kindling for nightmarish debugging sessions. If the double negative is a problem to you, feel free to swap what's in the `if` block and what's outside.

**Comment by gballet** [comment]
> @gballet this change is related to golang code generated from an ABI definition. It has nothing to do with clef and is totally separate functionality.

That's precisely what I'm asking you: why should we duplicate functionalities and therefore the maintenance effort when those functionalities are provided by dedicated modules?

> As for testing, as far as I am aware there are no tests at all for this file so there is no base on which to build

https://golang.org/pkg/testing/

**Comment by mcdee** [comment]
@gballet `clef` is a command-line interface to sign arbitrary transactions, `abigen` generates golang bindings for smart contracts.  They are totally different things.  If you believe there is any overlap please explain where you see it.

And no, I'm not going to write an entire test harness for the `abigen`-related code.  If someone writes the tests for the existing code I'll happily add the tests for the changes in this PR.

**Comment by mcdee** [comment]
> No overlap. Referring to your Jan 5 question "is there another way you propose to obtain a signed but unsent transaction?" I provided you one: using clef. I want to understand why that wouldn't work for you and would justify adding a feature that can be realized with already-existing tools.

I already said this patch is related to the golang bindings generated by `abigen`.  Clef does nothing with ABIs and has no golang bindings, so it can neither generate the transaction from the contract method nor can transactions be generated or signed from within a golang program.

**Comment by gballet** [comment]
You don't need go bindings for clef and generating the transactions can be done easily within go by using the abi package. Anyhow, let's drop the matter in the spirit of Christmas or something like that.

Do you still have comments about* the test and rewrite points?

**Pull request closed without merging**